design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,350 @@
+
+These need to be defined for each authenticator.
+To simplify that, we can define a simple class for each of these constraints.
+The whole constraints will be an array of these constraints, defined once for each authenticator.
+Then, to validate the Resource Restrictions, it only needs to loop on these constraints, and invoke
+`constraint.validate(resource_restrictions)` on each.
+
+This way, adding or changing the logic is more simple and intuitive, while maintaining readability.
+
+Each of the constraints classes will have the same `validate` signature, and will look similar to:
+
+```ruby
+class RequiredConstraint
+
+  def initialize(required:)
+    @required = required
+  end
+
+  def validate(resource_restrictions:)
+    # implement the constraint
+  end
+
+end
+```
+
+It will be used by each authenticator to define how to validate the configured
+restrictions. These are the ones that were defined in the policy for the host,
+regardless of the request.
+
+Example how it will look like for `authn-azure`:
+
+```ruby
+REQUIRED = [SUBSCRIPTION_ID, RESOURCE_GROUP]
+IDENTITY_EXCLUSIVE = [USER_ASSIGNED_IDENTITY, SYSTEM_ASSIGNED_IDENTITY]
+PERMITTED = REQUIRED + IDENTITY_EXCLUSIVE
+
+CONSTRAINTS =
+    [
+        RequiredConstraint.new(required: REQUIRED),
+        PermittedConstraint.new(permitted: PERMITTED),
+        ExclusiveConstraint.new(exclusive: IDENTITY_EXCLUSIVE)
+    ]
+```
+
+The general class will be defined as follows:
+
+- input
+  - An array of constraints, command classes that accept `resource_restrictions` as input.
+  - The `resource_restrictions` hash built in the previous step
+- output
+  - Nothing, relevant error raised if it failed
+
+So the class signature is:
+
+```ruby
+ValidateResourceRestrictions = CommandClass.new(
+  dependencies: {
+    logger: Rails.logger
+  },
+  inputs:   %i(resource_restrictions constraints)
+) do
+
+  def call
+    validate_restrictions_meet_constraints
+  end
+
+  private
+  
+  #...
+
+end
+```
+
+#### 3. Validate the Resource Restrictions match the authentication request
+
+Once we extracted the Resource Restrictions from the host's annotations, and
+verified it is defined correctly, we can validate that the authentication
+request is valid by validating the Resource Restrictions. For example, if
+the k8s host is permitted to authenticate from namespace `namespace-1`, and
+the authentication request was sent from a pod that is in namespace
+`namespace-2` we will fail the request.
+
+All we need to do is iterate over the Resource Restrictions hash 
+and verify that all of its fields equal the Authentication request object:
+
+```ruby
+resource_restrictions.all {|restriction, value| properties[restriction] == value}
+```
+
+So the general class will be defined as follows:
+- input
+  - Host Resource Restrictions object
+  - Request's Properties object
+- output
+  - Nothing, raise an error if failed
+  
+```ruby
+ValidateParametersMatchResourceRestrictions = CommandClass.new(
```

Comment:
not sure that `parameters` is very intuitive. How about `ValidateRequestMatchsResourceRestrictions`?

Reply:
How do you like it now?

Reply:
not really sure where the new one is but i'll see it in the implementation PR so don't bother.